Added valid_password method in ValidationRules

najdanovicivan opened the pull request:
Added method in validation rules to be able to check user current password. This is intended to be used in admin form for checks when user is modifying password.

The rule takes one parameter which is user ID so it can be used in form/model validation like this

valid_password[{id}]

michalsn commented:
I think that we should separate the code responsible for validating password to a separate method in the `LocalAuthenticator` class (right now it's a part of `validate()` method). 

Then we should use this method in validation rule since in the future `valid_password` implementation may depend on the used authentication library. So, I guess we shouldn't tightly couple this rule with `LocalAuthenticator`.

Also... if this rule is intended only for currently logged in users, why not skip the parameter part and use `user()` helper method to grab user data inside the rule?

najdanovicivan commented:
@michalsn You're right about not needing parameter so I've updated the code to get the user via auth helper user() method.

I've also added one additional rule valid_password_with which works in the same manner as required_with so I had this rules previously

```
permit_empty|required_with[newPassword]|valid_password[{id}]
```

Which can now be replaced with single rule

```
'valid_password_with[newPassword]'
```

I'like the idea of using LocalAuthenticator but splitting the password check will also require updating the AuthenticatorInterface. So it might be a tricky to do this properly. The only way that looks feasible to me is moving password check to method called 

validate_password($entity, $password)

And then use is_callable on Authenticator to check if validate_password exists as it does not make sense to have validate_password in the interface as it won't be possible to implement it in any password less authentication.

michalsn commented:
> I've also added one additional rule valid_password_with which works in the same manner as required_with so I had this rules previously. 

My personal preference is that if something can be done with simple 2 rules, then adding another one to do the same is unnecessary. Because we simply end up with an additional code that we have to maintain.

> I'like the idea of using LocalAuthenticator but splitting the password check will also require updating the AuthenticatorInterface. So it might be a tricky to do this properly.

Actually I can imagine `validate_password()` method in the `AuthenticatorInterface`. If this method were not supported by the "driver", then I guess it would have to return `false` (or even throw an exception) - simple as that. There is still no official release yet, so I guess we can do this kind of change?

Maybe Lonnie will have some opinion about it and idea how to handle this. Or maybe he will say to just leave it as it is :)

lonnieezell commented:
Sorry this one slipped by me. I agree with @michalsn  here. I'm good adding that to the interface, but don't think `valid_password_with` is needed.

lonnieezell commented:
@najdanovicivan Are you interested in making these changes or can I close?

najdanovicivan commented:
@lonnieezell I’m on it

najdanovicivan commented:
I've done the changes to include validate_password method in the AuthenticatorInterface but now I have some additional thoughts

What about renaming the method to validate_secret($user, $secret) that way it will be more generic so it will be useable for example with session tokens

 Now in valid password $authenticator is get with Service methods which have $authenticationLib set to `local` as default. 

It might be a good idea to change rule name to valid_secret as well and add a second optional parameter for the authenticationLib with default value of `local`

najdanovicivan commented:
@lonnieezell @MGatner  Can we see this one through so it can get merged.  I did not get any response related to renaming the method to validate credential

MGatner commented:
@lonnieezell will have to weigh in on the name change.

lonnieezell commented on src/Authentication/LocalAuthenticator.php:
```diff
@@ -186,4 +184,23 @@ public function validate(array $credentials, bool $returnUser=false)
+    public function validate_password(User $user, string $password) : bool
+    {
+        // Can't validate without a password.
+        if (empty($credentials['password']) || count($credentials) < 2)
```
There's a few problems here: 

1. The `$credentials` array doesn't exist in this method
2. I prefer to  return early instead of wrapping code in an if statement like this. 
3. Even if all of the other stuff was good, the current implementation doesn't always return a boolean like the method signature states.

lonnieezell commented on src/Authentication/Passwords/ValidationRules.php:
```diff
@@ -60,6 +60,27 @@ public function strong_password(string $value, string &$error1 = null, array $da
     }
 
+    /**
+     * A validation helper method to check if the passed
```
```suggestion
     * A validation helper method to check if the 
```

lonnieezell commented on src/Authentication/LocalAuthenticator.php:
```diff
@@ -186,4 +184,23 @@ public function validate(array $credentials, bool $returnUser=false)
+     *
+     * @return bool
+     */
+    public function validate_password(User $user, string $password) : bool
```
Typically all method names are camel-cased unless used for validation rules. 

```suggestion
    public function validatePassword(User $user, string $password) : bool
```

lonnieezell reviewed (changes requested):
Looking this over more during my review and I think this is a little too limited of a use-case to be included in the core library. 

Thanks for your work on it, though, and hopefully you'll consider making other PR's in the future.
